### ISSUE TITLE
Support custom error messages for optionality via nullability feature

### ISSUE THREAD

**Issue by captainsafia** [dotnet/aspnetcore#35501]
This issue captures user feedback from https://twitter.com/davidfowl/status/1428228518953984005?s=20.

Currently, the feature to support treating parameters that are nullable or have default values as optional will return a status code and no error message as part of the response. This is largely because (1) it builds on top of the existing try-parse parsing logic which does nothing and (2) when we were working through the feature we had punted on the problem of what message to send back.

Users want to be be able to customize the message to:
- Provide clearer info to the client
- Conform to a specific API implementation

**Comment by davidfowl** [comment]
I think this goes beyond the optionality feature. It's anytime the framework automagically returns a response (e.g. if TryParse fails).

**Comment by captainsafia** [comment]
> I think this goes beyond the optionality feature. It's anytime the framework automagically returns a response (e.g. if TryParse fails).

True, true. The fact that both the TryParse and optionality checks both build on top of the `CheckParams` logic means that they both provide status codes with no messages to the user.

**Comment by DamianEdwards** [comment]
Folks can customize the message via a middleware, using a marker object in metadata or items to enable/disable per endpoint (inspired by [this reply tweet](https://twitter.com/SilvairLSoares/status/1428345482191978510?s=20))

![image](https://user-images.githubusercontent.com/249088/130109128-9775b5d7-5ffb-44d1-8ac9-58c4df6a5603.png)

**Comment by davidfowl** [comment]
What's nice about this is that we don't need to add anymore configuration or concepts. It's just middleware.

**Comment by poke** [comment]
It would still be good if there was some global way to affect this. If the validating behavior is there by default with non-nullable parameters, then there should also be some way to control this default behavior.

I could imagine some callback that can be registered via options somewhere which default value is just the current 400 response behavior?

**Comment by DamianEdwards** [comment]
We literally have no extensibility points like that today in Minimal APIs and have been very hesitant to do so for .NET 6. We want to let people use and see what the wider feedback is before we start committing to API and more layers of extensibility (complexity). One of the reasons MVC is so complex today is because so much is customizable. While powerful it makes it difficult to learn and limits what we can do regarding performance improvements.

If it can be done with middleware and other existing concepts that don't require new API in a reasonable way, we intend to leave it that way, for now.

**Comment by poke** [comment]
> While powerful it makes it difficult to learn

If “learning” or discoverability is a goal of the new minimal API, then I don’t think it should automatically respond with a 400 if it wasn’t actually produced in user code. Because that really is undiscoverable magic from the framework.

Especially if it returns just a blank 400 without any details, then that really makes it very difficult to diagnose and learn about this too. I deal often enough with users that aren’t even aware of how to access their server logs, who mostly rely on the content of the HTTP responses for figuring out what’s wrong. If there’s now a blank response, then imo this will do more harm than good.

And if you decide to produce some content, e.g. using ProblemDetails, then there _needs_ to be a way to reconfigure this since ProblemDetails, being a relatively new spec, is only one of many ways to encode errors. And you wouldn’t want to have the framework be limiting here, especially since the minimal API is mostly a “do whatever you want, manually” way.

It would probably better to just throw an exception then which people could catch in a custom middleware to do custom handling, or which would bubble up until some default thing handles it.

**Comment by davidfowl** [comment]
Maybe a decent tradeoff could be using a response header. That would: 
- Put something in the response by default
- Allow users to customize the response with above approach

We'd also continue logging what happened.

**Comment by DamianEdwards** [comment]
@davidfowl by that you mean the framework would set a header by default? What would it be (name/value)?

**Comment by davidfowl** [comment]
> @davidfowl by that you mean the framework would set a header by default? What would it be (name/value)?

Yea, maybe just the exception text.

**Comment by DamianEdwards** [comment]
So a custom header with the exception message? Which exceptions are we talking about as I assume it would be constrained to some specific scenarios?

**Comment by davidfowl** [comment]
Or the problem details message?

**Comment by DamianEdwards** [comment]
OK so this is pretty gnarly I admit but I added [some examples to my playground repo](https://github.com/DamianEdwards/MinimalApiPlayground/blob/84446f016b7bc7f3a2b91a5ef111933899eba749/src/MinimalApiPlayground/Program.cs#L180) (rc.2 branch) that use a few tricks to enable suppressing the default behavior (binding and/or response defaults).

The magic is basically a helper type [`DefaultBinder<TValue>`](https://github.com/DamianEdwards/MinimalApiPlayground/blob/84446f016b7bc7f3a2b91a5ef111933899eba749/src/MinimalApiPlayground/Properties/DefaultBinderOfT.cs) that allows you to invoke the framework's default logic explicitly, returning you the `TValue` value and the status code the framework **would have** replied with. It's not completely fool-proof but goes a pretty long way. It has to use ref emit right now to generate the dummy route handler delegates to invoke the binding but I imagine in a future version of the framework we could actually include a type like this.

It means you can write route handlers like this where the logic to handle any issues is in the handler itself:
``` c#
app.MapPost("/suppress-defaults", (SuppressDefaultResponse<Todo?> todo, HttpContext httpContext) =>
    {
        if (todo.Exception != null)
        {
            // There was an exception during binding, handle it however you like
            throw todo.Exception;
        }

        if (todo.StatusCode != 200)
        {
            // The default logic would have auto-responded, do what you like instead
            throw new BadHttpRequestException("Your request was bad and you should feel bad", todo.StatusCode);
        }

        return Results.Ok(todo.Value);
    })
    .WithTags("Examples")
    .Accepts<Todo>("application/json");
```

**Comment by DamianEdwards** [comment]
For .NET 6 there won't be any first-class feature/option for changing the behavior including status codes and messages. However as noted above one can use middleware or a custom implementation of `BindAsync` to affect this behavior and ultimately control the response. We'll look at potential framework changes in .NET 7 based on how we see this being used.

**Comment by DamianEdwards** [comment]
This is something we should revisit as part of the endpoint filters feature, i.e. it should be possible to author a filter that runs before the default logic and return a custom error message.

**Comment by captainsafia** [comment]
We have this set to a priority 3. We should re-evaluate since this is coming in a few discussions.

**Comment by PaulusParssinen** [comment]
Having to do custom BindAsync (or any other trickery around ASP.NET Core internals) is really discouraging here. Would really like to see this prioritized again with the other Minimal API validation improvement effort.

**Comment by PaulusParssinen** [comment]
This is one of the biggest issues which weighs on my mind whenever someone asks me whether to use Minimal APIs or not. There's no way to give proper response to public API consumer why any primitive type binding fails with RDG-based approach. You can hit this with simply accepting a basic date/time primitive as your request/route parameter resulting nothing other than empty BadRequest response for slightly malformed input causing `IParsable.TryParse` return `false`.

It really makes me seriously consider the option of going back to controllers which I hate.

**Comment by cmerighi** [comment]
Still nothing on this?
It should have been a core concern from day one of DataAnnotations-based validation in minimal APIs...
This issue is 5yo.
